sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -180,9 +180,9 @@ class Dataset[T] private[sql](
     // to happen right away to let these side effects take place eagerly.
     queryExecution.analyzed match {
       case c: Command =>
-        LocalRelation(c.output, queryExecution.executedPlan.executeCollect())
+        LocalRelation(c.output, withAction("collect", queryExecution)(_.executeCollect()))
       case u @ Union(children) if children.forall(_.isInstanceOf[Command]) =>
-        LocalRelation(u.output, queryExecution.executedPlan.executeCollect())
+        LocalRelation(u.output, withAction("collect", queryExecution)(_.executeCollect()))
```

Comment:
shall we only add execution id for commands that will trigger execution? AFAIK there are 3 commands: `CreateDataSourceTableAsSelectCommand`, `CreateHiveTableAsSelectCommand` and `CacheTable`. We can call `SQLExecution.withNewExecutionId` inside these 3 commands.

Then we don't need to worry about nested execution.